### PR TITLE
do_balance internals significantly DRYed out, orchard notes and balance now stored in wallet

### PR DESCRIPTION
**Pull request by AloeareV**

### CODE REVIEW

File: lib/src/lightclient.rs [hunk 1]
```diff
@@ -589,19 +600,25 @@ impl LightClient {
...
-            "spendable_zbalance" => self.wallet.spendable_zbalance(None).await,
-            "unverified_zbalance"   => self.wallet.unverified_zbalance(None).await,
-            "tbalance"           => self.wallet.tbalance(None).await,
-            "z_addresses"        => z_addresses,
-            "t_addresses"        => t_addresses,
+            "sapling_balance"                 => self.wallet.sapling_balance(None).await,
+            "verified_sapling_balance"        => self.wallet.verified_sapling_balance(None).await,
+            "spendable_sapling_balance"       => self.wallet.spendable_sapling_balance(None).await,
+            "unverified_sapling_balance"      => self.wallet.unverified_sapling_balance(None).await,
+                "orchard_balance"             => self.wallet.orchard_balance(None).await,
```
Comment (zancas) [inline lib/src/lightclient.rs:614]: odd indentation...

File: lib/src/lightclient.rs [hunk 2]
```diff
@@ -589,19 +600,25 @@ impl LightClient {
...
+        // TODO:  Think about exposing an interface that references the Orchard Pool.
         object! {
-            "zbalance"           => self.wallet.zbalance(None).await,
-            "verified_zbalance"  => self.wallet.verified_zbalance(None).await,
-            "spendable_zbalance" => self.wallet.spendable_zbalance(None).await,
-            "unverified_zbalance"   => self.wallet.unverified_zbalance(None).await,
-            "tbalance"           => self.wallet.tbalance(None).await,
-            "z_addresses"        => z_addresses,
-            "t_addresses"        => t_addresses,
+            "sapling_balance"                 => self.wallet.sapling_balance(None).await,
```
Comment (zancas) [inline lib/src/lightclient.rs:610]: Almost repetitive enough for a declarative macro?

File: lib/src/wallet/traits.rs [hunk 3]
```diff
@@ -143,19 +144,21 @@ impl FromBytes<32> for OrchardNullifier {
...
+    Self: Sized,
+{
+    fn from_commitment(from: &[u8; 32]) -> CtOption<Self>;
 }
 
 impl FromCommitment for SaplingNode {
-    fn from_commitment(from: &[u8; 32]) -> Self {
-        Self::new(*from)
+    fn from_commitment(from: &[u8; 32]) -> CtOption<Self> {
+        CtOption::new(Self::new(*from), subtle::Choice::from(1))
```
Comment (zancas) [inline lib/src/wallet/traits.rs:156]: I don't understand the motivation here.

File: lib/src/wallet/traits.rs [hunk 4]
```diff
@@ -383,13 +398,100 @@ impl NoteAndMetadata for SaplingNoteAndMetadata {
...
     fn wallet_transaction_notes_mut(wallet_transaction: &mut WalletTx) -> &mut Vec<Self> {
         &mut wallet_transaction.sapling_notes
     }
 
+    fn wallet_transaction_notes(wallet_transaction: &WalletTx) -> &Vec<Self> {
+        &wallet_transaction.sapling_notes
+    }
+}
+
+impl NoteAndMetadata for OrchardNoteAndMetadata {
```
Comment (zancas) [inline lib/src/wallet/traits.rs:422]: Wait..  wasn't this already in place..  why does it seem to be added in this PR?

File: lib/src/wallet/traits.rs [hunk 5]
```diff
@@ -497,65 +599,6 @@ impl WalletKey for OrchardKey {
     }
 }
 
-impl NoteAndMetadata for OrchardNoteAndMetadata {
```
Comment (zancas) [inline lib/src/wallet/traits.rs:500]: Ohh...  it's moved.  I see.

File: lib/src/wallet/traits.rs [hunk 6]
```diff
@@ -608,3 +651,37 @@ impl<P: Parameters> DomainWalletExt<P> for OrchardDomain {
         &mut transaction.orchard_notes
     }
 }
+
+pub(crate) trait Diversifiable {
```
Comment (zancas) [inline lib/src/wallet/traits.rs:655]: Why does this look like it's being added for the first time?

File: lib/src/lightclient.rs (hunk 2 above)
Comment (AloeareV) [reply lib/src/lightclient.rs:617]: Hmm.. That's a good thought. Issued.

File: lib/src/wallet/traits.rs (hunk 4 above)
Comment (AloeareV) [reply lib/src/wallet/traits.rs:432]: This is moved from elsewhere in the same file, I'm pretty sure.

File: lib/src/wallet/traits.rs (hunk 6 above)
Comment (AloeareV) [reply lib/src/wallet/traits.rs:672]: Well.. Because I think it is!

File: lib/src/lightclient.rs (hunk 1 above)
Comment (AloeareV) [reply lib/src/lightclient.rs]: Fixed!

File: lib/src/blaze/block_witness_data.rs [hunk 7]
```diff
@@ -3,13 +3,13 @@ use crate::{
     grpc_connector::GrpcConnector,
     lightclient::checkpoints::get_all_main_checkpoints,
     wallet::{
-        data::{BlockData, WalletNullifier, WalletTx, WitnessCache},
-        traits::FromCommitment,
-        transactions::WalletTxns,
+        data::{BlockData, TransactionMetadata, WalletNullifier, WitnessCache},
```
Comment (zancas) [inline lib/src/blaze/block_witness_data.rs:6]: `WalletNullifier` --> `ChannelNullifier`

File: lib/src/blaze/block_witness_data.rs [hunk 8]
```diff
@@ -523,8 +519,10 @@ impl BlockAndWitnessData {
                 CommitmentTree::empty()
             } else {
                 let tree_state = GrpcConnector::get_sapling_tree(uri, prev_height).await?;
```
Comment (zancas) [inline lib/src/blaze/block_witness_data.rs:521]: This `get_sapling_tree` fn is inappropriately specific....

File: lib/src/wallet.rs [hunk 9]
```diff
@@ -773,28 +774,51 @@ impl LightWallet {
...
+                    NnMd::transaction_metadata_notes(transaction)
+                        .iter()
+                        .filter(|nd| match addr.as_ref() {
+                            Some(a) => {
+                                use self::traits::Recipient as _;
+                                let diversified_address =
+                                    &nd.fvk().diversified_address(*nd.diversifier()).unwrap();
+                                *a == diversified_address.b32encode_for_network(&self.config.chain)
+                            }
+                            None => true,
```
Comment (zancas) [inline lib/src/wallet.rs:811]: Should this be `false`?

Comment (AloeareV) [reply lib/src/wallet.rs:811]: If we supply an address we're asking for the balance of that address. If we don't supply an address, we're asking for the balance of all addresses.

**Review by zancas — APPROVED** [review]

**Pull request merged**